src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,270 @@
+    this.threshold_ = this.threshold_.sort();
+    dev().assert(this.threshold_[0] >= 0 &&
+        this.threshold_[this.threshold_.length - 1] <= 1,
+        'Threshold should only contain value range from 0 to 1');
+
+    /** @private {?Element} */
+    this.element_ = null;
+
+    /** @private {boolean} */
+    this.isAmpElement_ = false;
+
+    /**
+     * A cached value to store getBoundingClientRect() value
+     * if element is not an AMP element
+     * @private {?DOMRect}
+     */
+    this.elementRectForNonAMP_ = null;
+
+    /**
+     * The prev threshold slot which the previous ratio fills
+     * Range [0, this.threshold_.length]
+     * TODO: Do we always want to call callback at the first tick?
+     * @private {number}
+     */
+    this.prevThresholdSlot_ = 0;
+
+    // TODO: create the PositionObserver class, or create listener for position
+    // info if inside an iframe
+
+    // TODO: Add the postMessageApi logic outside IntersectionObserverPolyfill.
+  }
+
+  /**
+   * Provide a way to observer the intersection change for a specific element
+   * Please note that the IntersectionObserverPolyfill only allow to observe one
+   * element at a time.
+   * TODO: Support observing multiple elements.
+   * @param {!Element} element
+   */
+  observe(element) {
+    // Reset cached value for a new element
+    this.isAmpElement_ = false;
+    this.elementRectForNonAMP_ = null;
+    this.prevThresholdSlot_ = 0;
+    // TODO: Need a good way to check if an element is AMP element
+    if (element.isBuilt && element.isBuilt()) {
+      this.isAmpElement_ = true;
+    }
+    this.element_ = element;
+  }
+
+  /**
+   * Tick function that update the DOMRect of the root of observed element.
+   * Caller needs to make sure to pass in the correct container.
+   * @param {!./layout-rect.LayoutRectDef} viewport.
+   * @param {./layout-rect.LayoutRectDef=} opt_container. relative to viewport
+   */
+  tick(viewport, opt_container) {
+    // Normalize container LayoutRect to be relative to page
+    if (opt_container) {
+      opt_container =
+          moveLayoutRect(opt_container, viewport.left, viewport.top);
+    }
+    let elementRect;
+    let ownerRect = null;
+
+    if (this.isAmpElement_) {
+      // If calls with container, always assume getLayoutBox() return relative
+      // LayoutRect to container
+      elementRect = this.element_.getLayoutBox();
+      ownerRect = this.element_.getOwnerLayoutBox();
+      if (opt_container) {
+        elementRect = moveLayoutRect(elementRect, opt_container.left,
+            opt_container.top);
+        if (ownerRect) {
+          ownerRect = moveLayoutRect(ownerRect, opt_container.left,
+              opt_container.top);
+        }
+      }
+    } else {
+      // If the element is not an AMP element, always assume its position to
+      // the page won't change.
+      if (!this.elementRectForNonAMP_) {
+        elementRect = this.element_./*OK*/getBoundingClientRect();
```

Comment:
This is relative to the viewport, which is not true for `BaseElement#getLayoutBox`. You have to add the `scrollTop` and `scrollLeft` to this value.